File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_time/TimePickerDialogBuilder.java
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_time;

import com.github.rooneyandshadows.lightbulb.dialogs.base.LightBulbDialogBuilder;
import com.github.rooneyandshadows.lightbulb.dialogs.base.LightBulbDialogFragment;
import com.github.rooneyandshadows.lightbulb.dialogs.base.LightBulbPickerDialogFragment;

import androidx.fragment.app.FragmentManager;

public class TimePickerDialogBuilder extends LightBulbDialogBuilder<TimePickerDialog> {

    private LightBulbPickerDialogFragment.SelectionChangedListener<int[]> timeSetListener;
    private int[] initialTime;

    public TimePickerDialogBuilder(FragmentManager manager, String dialogTag) {
        super(manager, dialogTag);
    }

    @Override
    public TimePickerDialogBuilder withTitle(String title) {
        return (TimePickerDialogBuilder) super.withTitle(title);
    }

    @Override
    public TimePickerDialogBuilder withMessage(String message) {
        return (TimePickerDialogBuilder) super.withMessage(message);
    }

    @Override
    public TimePickerDialogBuilder withPositiveButton(LightBulbDialogFragment.DialogButtonConfiguration positiveButtonConfiguration, LightBulbDialogFragment.DialogButtonClickListener onClickListener) {
        return (TimePickerDialogBuilder) super.withPositiveButton(positiveButtonConfiguration, onClickListener);
    }

    @Override
    public TimePickerDialogBuilder withNegativeButton(LightBulbDialogFragment.DialogButtonConfiguration negativeButtonConfiguration, LightBulbDialogFragment.DialogButtonClickListener onClickListener) {
        return (TimePickerDialogBuilder) super.withNegativeButton(negativeButtonConfiguration, onClickListener);
    }

    @Override
    public TimePickerDialogBuilder withOnCancelListener(LightBulbDialogFragment.DialogCancelListener listener) {
        return (TimePickerDialogBuilder) super.withOnCancelListener(listener);
    }

    @Override
    public TimePickerDialogBuilder withOnShowListener(LightBulbDialogFragment.DialogShowListener listener) {
        return (TimePickerDialogBuilder) super.withOnShowListener(listener);
    }

    @Override
    public TimePickerDialogBuilder withOnHideListener(LightBulbDialogFragment.DialogHideListener listener) {
        return (TimePickerDialogBuilder) super.withOnHideListener(listener);
    }

    @Override
    public TimePickerDialogBuilder withCancelOnClickOutsude(boolean closeOnClickOutside) {
        return (TimePickerDialogBuilder) super.withCancelOnClickOutsude(closeOnClickOutside);
    }

    @Override
    public TimePickerDialogBuilder withDialogType(LightBulbDialogFragment.DialogTypes dialogType) {
        return (TimePickerDialogBuilder) super.withDialogType(dialogType);
    }

    @Override
    public TimePickerDialogBuilder withAnimations(LightBulbDialogFragment.DialogAnimationTypes animation) {
        return (TimePickerDialogBuilder) super.withAnimations(animation);
    }

    public TimePickerDialogBuilder withOnDateSelectedEvent(LightBulbPickerDialogFragment.SelectionChangedListener<int[]> listener) {
        this.timeSetListener = listener;
        return this;
    }

    public TimePickerDialogBuilder withInitialTime(int hour, int minute) {
        this.initialTime = new int[]{hour, minute};
        return this;
    }

    public TimePickerDialogBuilder withInitialTime(int[] time) {
        this.initialTime = time;
        return this;
    }

    @Override
    public TimePickerDialog buildDialog() {
        TimePickerDialog dialogFragment = (TimePickerDialog) fragmentManager.findFragmentByTag(dialogTag);
        if (dialogFragment == null)
            dialogFragment = TimePickerDialog.newInstance(positiveButtonConfiguration, negativeButtonConfiguration, cancelableOnClickOutside, animation);
        dialogFragment.setFragmentManager(fragmentManager);
        dialogFragment.setDialogTag(dialogTag);
        dialogFragment.addOnNegativeClickListeners(onNegativeClickListener);
        dialogFragment.addOnPositiveClickListener(onPositiveClickListener);
        dialogFragment.setSelection(initialTime);
        dialogFragment.addOnShowListener(onShowListener);
        dialogFragment.addOnHideListener(onHideListener);
        dialogFragment.addOnCancelListener(onCancelListener);
        dialogFragment.setOnSelectionChangedListener(timeSetListener);
        return dialogFragment;
    }
}

File: lightbulb-dialogs/src/main/java/com/github/rooneyandshadows/lightbulb/dialogs/picker_dialog_time/TimePickerDialogBuilder.kt
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_time

import android.os.Bundle
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.LifecycleOwner
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment.SelectionChangedListener
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.*
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.callbacks.*
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_time.TimePickerDialog.Time

@Suppress("unused")
class TimePickerDialogBuilder @JvmOverloads constructor(
    lifecycleOwner: LifecycleOwner? = null,
    manager: FragmentManager,
    dialogTag: String,
) : BaseDialogBuilder<TimePickerDialog>(lifecycleOwner, manager, dialogTag) {
    private var timeSetListener: SelectionChangedListener<Time>? = null
    private var initialTime: Time? = null

    @Override
    override fun setupNonRetainableSettings(dialog: TimePickerDialog) {
        dialog.apply {
            timeSetListener?.apply { addOnSelectionChangedListener(this) }
        }
    }

    @Override
    override fun setupRetainableSettings(dialog: TimePickerDialog) {
        dialog.apply {
            setSelection(initialTime)
        }
    }

    @Override
    override fun initializeNewDialog(): TimePickerDialog {
        return TimePickerDialog.newInstance()
    }

    @Override
    override fun withInitialDialogState(savedState: Bundle?): TimePickerDialogBuilder {
        return super.withInitialDialogState(savedState) as TimePickerDialogBuilder
    }

    @Override
    override fun withTitle(title: String): TimePickerDialogBuilder {
        return super.withTitle(title) as TimePickerDialogBuilder
    }

    @Override
    override fun withMessage(message: String): TimePickerDialogBuilder {
        return super.withMessage(message) as TimePickerDialogBuilder
    }

    @Override
    override fun withButton(configuration: DialogButton): TimePickerDialogBuilder {
        return super.withButton(configuration) as TimePickerDialogBuilder
    }

    @Override
    override fun withOnCancelListener(listener: DialogCancelListener): TimePickerDialogBuilder {
        return super.withOnCancelListener(listener) as TimePickerDialogBuilder
    }

    @Override
    override fun withOnShowListener(listener: DialogShowListener): TimePickerDialogBuilder {
        return super.withOnShowListener(listener) as TimePickerDialogBuilder
    }

    @Override
    override fun withOnHideListener(listener: DialogHideListener): TimePickerDialogBuilder {
        return super.withOnHideListener(listener) as TimePickerDialogBuilder
    }

    @Override
    override fun withCancelOnClickOutside(closeOnClickOutside: Boolean): TimePickerDialogBuilder {
        return super.withCancelOnClickOutside(closeOnClickOutside) as TimePickerDialogBuilder
    }

    @Override
    override fun withDialogType(dialogType: DialogTypes): TimePickerDialogBuilder {
        return super.withDialogType(dialogType) as TimePickerDialogBuilder
    }

    @Override
    override fun withAnimations(animation: DialogAnimationTypes): TimePickerDialogBuilder {
        return super.withAnimations(animation) as TimePickerDialogBuilder
    }

    @Override
    override fun withDialogListeners(listeners: DialogListeners): TimePickerDialogBuilder {
        return super.withDialogListeners(listeners) as TimePickerDialogBuilder
    }

    @Override
    fun withOnDateSelectedEvent(listener: SelectionChangedListener<Time>): TimePickerDialogBuilder {
        timeSetListener = listener
        return this
    }

    @Override
    fun withInitialTime(hour: Int, minute: Int): TimePickerDialogBuilder {
        initialTime = Time(hour, minute)
        return this
    }

    @Override
    fun withInitialTime(time: Time): TimePickerDialogBuilder {
        initialTime = time
        return this
    }
}